lib/iopromise/data_loader.rb

```diff
@@ -4,11 +4,30 @@ module IOPromise
   module DataLoader
     module ClassMethods
       def attr_promised_data(*args)
+        build_func = if args.last&.is_a?(Proc)
+          args.pop
+        else
+          nil
+        end
+
         @promised_data_keys ||= []
         @promised_data_keys.concat(args)
 
         args.each do |arg|
-          self.class_eval("def #{arg};@#{arg}.sync;end")
+          if build_func.nil?
+            self.class_eval("def async_#{arg};@#{arg};end")
+          else
+            self.define_method("async_#{arg}") do
+              @promised_data_memo ||= {}
+              @promised_data_memo[arg] ||= if build_func.arity == 1
+                self.instance_exec(arg, &build_func)
```

Comment:
I went down this path only because it matched `attr_reader`, and it made sense when it only worked against an instance variable just like `attr_reader`. But now that we're adding lambdas, I agree that that ends up more confusing.

:+1: Let's simplify this to be just one attribute and an optional lambda - it's a subset of the current functionality and easier to maintain, and we can always extend it later if we come across a compelling case for it. In the meantime, it makes everything easier to reason about.